OpenSearch Ingestion: new resources pipeline endpoint and resource policy resources

graytaylor0 opened the pull request:
<!---
See what makes a good Pull Request at: https://hashicorp.github.io/terraform-provider-aws/raising-a-pull-request/
--->

<!-- heimdall_github_prtemplate:grc-pci_dss-2024-01-05 -->

## Rollback Plan

If a change needs to be reverted, we will publish an updated version of the library.

This contains new resources that will not break any existing functionality for terraform support for osis service

## Changes to Security Controls

Are there any changes to security controls (access controls, encryption, logging) in this pull request? If so, explain.

### Description
<!---
Please provide a helpful description of what change this pull request will introduce.
--->

This change adds support for 2 new resources within OpenSearch Ingestion (osis) service

1. Pipeline endpoints
2. Pipeline resource policies 

(https://docs.aws.amazon.com/opensearch-service/latest/developerguide/cross-account-pipelines.html)


### Relations
<!---
If your pull request fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates.

For Example:

Relates #0000
or 
Closes #0000
--->

Closes #0000

### References
<!---
Optionally, provide any helpful references that may help the reviewer(s).
--->

https://docs.aws.amazon.com/opensearch-service/latest/developerguide/cross-account-pipelines.html



### Output from Acceptance Testing
<!--
Replace TestAccXXX with a pattern that matches the tests affected by this PR.

Replace ec2 with the service package corresponding to your tests.

For more information on the `-run` flag, see the `go test` documentation at https://tip.golang.org/cmd/go/#hdr-Testing_flags.
-->

```console
% make testacc TESTS=TestAccOSISPipelineEndpoint_basic
=== RUN   TestAccOSISPipelineEndpoint_basic
Pipeline endpoint name: "tf-acc-test-dgl13um4nb"
=== PAUSE TestAccOSISPipelineEndpoint_basic
=== CONT  TestAccOSISPipelineEndpoint_basic
PreCheck completed
--- PASS: TestAccOSISPipelineEndpoint_basic (803.62s)
PASS

% make testacc TESTS=TestAccOSISPipelineResourcePolicy_basic
=== RUN   TestAccOSISPipelineResourcePolicy_basic
=== PAUSE TestAccOSISPipelineResourcePolicy_basic
=== CONT  TestAccOSISPipelineResourcePolicy_basic
--- PASS: TestAccOSISPipelineResourcePolicy_basic (676.74s)
PASS
...
```

graytaylor0 committed:
Add provider for pipeline endpoints and resource policy resources and APIs

Signed-off-by: Taylor Gray <tylgry@amazon.com>

graytaylor0 commented on internal/service/osis/pipeline_resource_policy.go:
```diff
@@ -0,0 +1,180 @@
+type pipelineResourcePolicyResourceModel struct {
+	framework.WithRegionModel
+	ID          types.String      `tfsdk:"id"`
+	ResourceARN fwtypes.ARN       `tfsdk:"resource_arn"`
```
It is the same convention used for pipeline (https://github.com/hashicorp/terraform-provider-aws/blob/f5ed7f2de5a951530edf5ddbc01a2cc5e30883ce/internal/service/osis/pipeline.go#L490).

graytaylor0 commented on internal/service/osis/pipeline_endpoint.go:
```diff
@@ -0,0 +1,335 @@
+type pipelineEndpointResourceModel struct {
+	framework.WithRegionModel
+	ID          types.String                                                     `tfsdk:"id"`
+	PipelineARN fwtypes.ARN                                                      `tfsdk:"pipeline_arn"`
```
The same convention is used for pipeline.go

dlvenable reviewed (approved):
Thanks @graytaylor0 for this change!

YakDriver committed:
Fix CheckFrameworkResourceDisappears argument in osis pipeline_resource_policy_test

YakDriver committed:
Code fixes, cleanup

YakDriver committed:
Fix OpenSearch service naming: brand to Amazon, OSIS human_friendly

YakDriver committed:
Update names

YakDriver committed:
Fix OSIS docs: subcategory and add pipeline_endpoint/resource_policy docs

YakDriver committed:
Add optional region argument to OSIS resource docs

YakDriver committed:
Add changelog for aws_osis_pipeline_endpoint and aws_osis_pipeline_resource_policy

YakDriver commented:
I have made some simple, driveby fixes. Going forward, this will still need additional work and review. The biggest improvement needed is that both resources should support resource identity.

YakDriver commented on internal/service/osis/pipeline_endpoint.go:
```diff
@@ -0,0 +1,335 @@
+	r := &pipelineEndpointResource{}
+
+	r.SetDefaultCreateTimeout(45 * time.Minute)
+	r.SetDefaultUpdateTimeout(45 * time.Minute)
```
Update is a noop so no timeout needed.

YakDriver commented on internal/service/osis/pipeline_endpoint.go:
```diff
@@ -0,0 +1,335 @@
+	if response.Diagnostics.HasError() {
+		return
+	}
+	data.VPCOptions = vpcOptions
```
This and line 127 are a little strange. I think what's happening is VPC options aren't returned so we're just setting what was there before. If this is correct, that's fine but could use a comment to explain.

graytaylor0 replied:
That is correct. I will add a comment

gdavison replied:
Here and in the `Read` function, can `endpoint.VpcOptions` not be used?

YakDriver commented on internal/service/osis/pipeline_endpoint.go:
```diff
@@ -0,0 +1,335 @@
+				},
+			},
+			names.AttrStatus: schema.StringAttribute{
+				Computed: true,
```
```suggestion
				Computed: true,
				PlanModifiers: []planmodifier.String{
					stringplanmodifier.UseStateForUnknown(),
				},
```

YakDriver commented on internal/service/osis/pipeline_endpoint.go:
```diff
@@ -0,0 +1,335 @@
+				Computed: true,
+			},
+			names.AttrVPCID: schema.StringAttribute{
+				Computed: true,
```
```suggestion
				Computed: true,
				PlanModifiers: []planmodifier.String{
					stringplanmodifier.UseStateForUnknown(),
				},
```

YakDriver commented on internal/service/osis/pipeline_endpoint_test.go:
```diff
@@ -0,0 +1,223 @@
+`, rName))
+}
+
+func testAccPipelineEndpointConfig_vpc(rName string) string {
```
```suggestion
func testAccPipelineEndpointConfig_pipeline(rName string) string {
```

The naming is confusing since the important thing from this config is the pipeline not the vpc.

YakDriver reviewed (commented):
I didn't review pipeline_resource_policy but here are some comments to chomp on until this is scheduled for a proper review.

graytaylor0 committed:
Update internal/service/osis/pipeline_endpoint.go

Co-authored-by: Dirk Avery <31492422+YakDriver@users.noreply.github.com>

graytaylor0 committed:
Update internal/service/osis/pipeline_endpoint_test.go

Co-authored-by: Dirk Avery <31492422+YakDriver@users.noreply.github.com>

graytaylor0 committed:
Update internal/service/osis/pipeline_endpoint.go

Co-authored-by: Dirk Avery <31492422+YakDriver@users.noreply.github.com>

graytaylor0 committed:
Update internal/service/osis/pipeline_endpoint.go

Co-authored-by: Dirk Avery <31492422+YakDriver@users.noreply.github.com>

graytaylor0 committed:
Update internal/service/osis/pipeline_endpoint_test.go

Co-authored-by: Dirk Avery <31492422+YakDriver@users.noreply.github.com>

graytaylor0 committed:
Update internal/service/osis/pipeline_endpoint.go

Co-authored-by: Dirk Avery <31492422+YakDriver@users.noreply.github.com>

graytaylor0 commented on internal/service/osis/pipeline_endpoint.go:
```diff
@@ -0,0 +1,335 @@
+	}
+
+	data.ID = fwflex.StringValueToFramework(ctx, aws.ToString(output.EndpointId))
+	data.EndpointID = fwflex.StringValueToFramework(ctx, aws.ToString(output.EndpointId))
```
Although the code is currently using the `EndpointId` in `findEndpointById`. I guess this can be replaced with the `ID`

graytaylor0 committed:
Remove uneccessary endpoint_id from pipeline endpoint resource

Signed-off-by: Taylor Gray <tylgry@amazon.com>

graytaylor0 committed:
Address PR comments for incorrect break and added comment explaning vpc options being set

Signed-off-by: Taylor Gray <tylgry@amazon.com>

gdavison commented on website/docs/cdktf/python/r/osis_pipeline.html.markdown:
The CDKTF documentation files can be removed

gdavison commented on internal/service/osis/pipeline_resource_policy.go:
```diff
@@ -0,0 +1,181 @@
+func (r *pipelineResourcePolicyResource) Schema(ctx context.Context, request resource.SchemaRequest, response *resource.SchemaResponse) {
+	response.Schema = schema.Schema{
+		Attributes: map[string]schema.Attribute{
+			names.AttrID: framework.IDAttribute(),
```
Because `id` is just duplicating `resource_arn` it should be removed.

gdavison replied:
This attribute, `id`, should be removed from the resource schema and model

gdavison commented on internal/service/osis/pipeline_resource_policy.go:
```diff
@@ -0,0 +1,181 @@
+	}
+}
+
+type pipelineResourcePolicyResource struct {
```
This declaration should be moved to directly after `newPipelineResourcePolicyResource`

gdavison commented on internal/service/osis/pipeline_resource_policy.go:
```diff
@@ -0,0 +1,181 @@
+
+// @FrameworkResource("aws_osis_pipeline_resource_policy", name="Pipeline Resource Policy")
+// @ArnIdentity("resource_arn", identityDuplicateAttributes="id")
+// @Testing(importIgnore="policy")
```
Why is this ignored?

gdavison commented on internal/service/osis/pipeline_resource_policy_test.go:
```diff
@@ -0,0 +1,211 @@
+
+func testAccPipelineResourcePolicyConfig_updated(rName string) string {
+	return acctest.ConfigCompose(fmt.Sprintf(testAccPipelineConfig_basic(rName), `
+
```
```suggestion
```

gdavison commented on internal/service/osis/pipeline_resource_policy_test.go:
```diff
@@ -0,0 +1,211 @@
+  resource_arn = aws_osis_pipeline.test.pipeline_arn
+
+  policy = <<EOF
+{
+  "Version": "2012-10-17",
+  "Id": "ingestPolicy",
+  "Statement": [{
+    "Sid": "AllowIngest",
+    "Effect": "Allow",
+    "Principal": {
+      "AWS": "${data.aws_caller_identity.target.account_id}"
+    },
+    "Action": [
+      "osis:CreatePipelineEndpoint"
+    ],
+    "Resource": "${aws_osis_pipeline.test.pipeline_arn}"
+  }]
+}
+EOF
```
This policy appears identical to the policy in `testAccPipelineResourcePolicyConfig_basic` and therefore won't trigger an update.

gdavison commented on internal/service/osis/pipeline_resource_policy_test.go:
```diff
@@ -0,0 +1,211 @@
+				Config: testAccPipelineResourcePolicyConfig_updated(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPipelineResourcePolicyExists(ctx, resourceName),
+				),
```
This should check that the value was changed to the expected new value. For example it could use `acctest.CheckResourceAttrJMES` or similar to verify a changed value in the JSON.

gdavison commented on internal/service/osis/pipeline_resource_policy_test.go:
```diff
@@ -0,0 +1,211 @@
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPipelineResourcePolicyExists(ctx, resourceName),
+				),
+			},
```
This should check that an update was triggered.


```suggestion
				ConfigPlanChecks: resource.ConfigPlanChecks{
					PostApplyPostRefresh: []plancheck.PlanCheck{
						plancheck.ExpectResourceAction(resourceName, plancheck.ResourceActionUpdate),
					},
				},
			},
```

gdavison commented on internal/service/osis/pipeline_endpoint_test.go:
```diff
@@ -0,0 +1,220 @@
+		},
+	})
+}
+
```
This needs a `_disappears` test

gdavison commented on internal/service/osis/pipeline_endpoint_test.go:
```diff
@@ -0,0 +1,220 @@
+  state = "available"
+}
+
+resource "aws_vpc" "endpoint_test" {
```
Because the resources in this section of the configuration are related to the resource that we're testing, they should be called `test`. The resources related to the source `aws_osis_pipeline` should be renamed to `pipeline` or something similar

gdavison commented on internal/service/osis/pipeline_endpoint_test.go:
```diff
@@ -0,0 +1,220 @@
+resource "aws_security_group" "endpoint_test" {
+  name   = %[1]q
+  vpc_id = aws_vpc.endpoint_test.id
+}
+
+resource "aws_osis_pipeline_endpoint" "test" {
+  pipeline_arn = aws_osis_pipeline.test.pipeline_arn
+
+  vpc_options {
+    subnet_ids         = [aws_subnet.endpoint_test.id]
+    security_group_ids = [aws_security_group.endpoint_test.id]
+  }
+}
```
Move the resource that we're testing to the top of the configuration for ease of reading

gdavison reviewed (changes requested):
Hi @graytaylor0, I have a number of suggested changes. The code doesn't currently compile, so I haven't been able to test it.

graytaylor0 commented:
@gdavison thanks for the review. Looks like some of my changes got removed on the rebase to address the merge conflicts. I will address these

graytaylor0 committed:
Address PR comments, fix compilation issues

Signed-off-by: Taylor Gray <tylgry@amazon.com>

graytaylor0 committed:
Remove redundant id from pipeline resource policy

Signed-off-by: Taylor Gray <tylgry@amazon.com>

graytaylor0 commented:
@gdavison I believe I have addressed all your comments. Let me know if any more changes are required. I verified it compiles and ran acceptance tests multiple times to confirm they are still passing. Thanks!

gdavison commented on internal/service/osis/pipeline_endpoint_test.go:
```diff
@@ -110,38 +138,38 @@ func testAccCheckPipelineEndpointExists(ctx context.Context, n string, v *awstyp
   tags = {
-    Name = %[1]q
+    Name = "%[1]s-endpoint"
   }
```
The `tags` isn't necessary here

gdavison commented on internal/service/osis/exports_test.go:
```diff
@@ -8,4 +8,10 @@ var (
+
+	ResourcePipelineEndpoint = newPipelineEndpointResource
+
+	ResourcePipelineResourcePolicy = newPipelineResourcePolicyResource
```
The `Find` references should be grouped together and the `Resource` references should be grouped together, with a single space between the groups

gdavison commented on internal/service/osis/pipeline_endpoint.go:
```diff
@@ -0,0 +1,331 @@
+}
+
+// FindPipelineEndpointByID retrieves a pipeline endpoint by its ID.
+func findPipelineEndpointByID(ctx context.Context, conn *osis.Client, endpointID string) (*awstypes.PipelineEndpoint, error) {
```
The capitalization of the documentation comment should match the function name

gdavison commented on internal/service/osis/pipeline_endpoint.go:
```diff
@@ -0,0 +1,331 @@
+	r := &pipelineEndpointResource{}
+
+	r.SetDefaultCreateTimeout(45 * time.Minute)
+	r.SetDefaultDeleteTimeout(45 * time.Minute)
```
Do these timeouts need to be this long?

graytaylor0 replied:
I can shorten them by looking at p99 for how long endpoints take to create and delete.

gdavison reviewed (changes requested):
Thanks for the updates, @graytaylor0. A few issues to address:

1. Some of the PR checks are failing. 
2. The tests `TestAccOpenSearchIngestionPipelineResourcePolicy_disappears` and `TestAccOpenSearchIngestionPipelineResourcePolicy_update` fail with `Attribute pipeline_name string length must be between 3 and 28, got: 31`.
3. When I run the acceptance tests, `TestAccOpenSearchIngestionPipelineEndpoint_basic` fails with `unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'. last error: Unable to create pipeline due to an internal exception.`. If I remove the explicit `availability_zone` from the `aws_subnets`, the tests pass. Are they required in your environment?
4. There were some unresolved comments from the last review and some new ones.

graytaylor0 committed:
Address additional comments and fix failing acceptance tests

Signed-off-by: Taylor Gray <tylgry@amazon.com>

gdavison commented on internal/service/osis/pipeline_resource_policy_test.go:
```diff
@@ -9,102 +9,113 @@ import (
 	ctx := acctest.Context(t)
-	rName := sdkacctest.RandomWithPrefix(acctest.ResourcePrefix)
+	rName := fmt.Sprintf("%s-%s", acctest.ResourcePrefix, acctest.RandString(t, 10))
 	resourceArn := "aws_osis_pipeline_resource_policy.test"
```
This should be renamed to `resourceName` like in the other tests

gdavison commented on internal/service/osis/pipeline_resource_policy_test.go:
```diff
@@ -0,0 +1,235 @@
+						return "", fmt.Errorf("Not found: %s", resourceName)
+					}
+					return rs.Primary.Attributes[names.AttrResourceARN], nil
+				},
```
This can use the existing function `acctest.AttrImportStateIdFunc`

```suggestion
				ImportStateIdFunc: acctest.AttrImportStateIdFunc(resourceName, names.AttrResourceARN),
```

gdavison commented on internal/service/osis/pipeline_resource_policy.go:
```diff
@@ -0,0 +1,175 @@
+		response.Diagnostics.Append(fwdiag.NewResourceNotFoundWarningDiagnostic(err))
+		response.State.RemoveResource(ctx)
+		return
+	}
```
`retry.NotFound` only checks for `retry.NotFoundError` or `sdkretry.NotFoundError`, which are internal to the provider, so this will never be `true`.

Since the unset or "deleted" state of the resource policy is `"{}"`, it would make sense to create a find function, e.g. `findPipelineResourcePolicyByResourceARN`, similar to `findPipelineEndpointByID`, which would treat the empty policy as `NotFound`. The find function could then also be used in the tests

gdavison commented on internal/service/osis/pipeline_resource_policy_test.go:
```diff
@@ -0,0 +1,235 @@
+				Config: testAccPipelineResourcePolicyConfig_basic(rName),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckPipelineResourcePolicyExists(ctx, t, resourceArn),
+					acctest.CheckFrameworkResourceDisappears(ctx, t, tfosis.ResourcePipelineResourcePolicy, resourceArn),
+				),
+				ExpectNonEmptyPlan: true,
+				ConfigPlanChecks: resource.ConfigPlanChecks{
+					PostApplyPostRefresh: []plancheck.PlanCheck{
+						plancheck.ExpectResourceAction(resourceArn, plancheck.ResourceActionCreate),
+					},
+				},
```
For completeness, this should check `PreApply` as well.

gdavison commented on internal/service/osis/pipeline_resource_policy_test.go:
```diff
@@ -0,0 +1,235 @@
+			if err != nil {
+				// If we get an error, the resource is gone
+				continue
+			}
```
Only a not found error should be treated as success

gdavison commented on internal/service/osis/pipeline_endpoint.go:
```diff
@@ -0,0 +1,324 @@
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	sdkretry "github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry"
```
Uses of `github.com/hashicorp/terraform-plugin-sdk/v2/helper/retry` should be replaced with `github.com/hashicorp/terraform-provider-aws/internal/retry`

gdavison commented on internal/service/osis/pipeline_resource_policy_test.go:
```diff
@@ -0,0 +1,235 @@
+	})
+}
+
+func TestAccOpenSearchIngestionPipelineResourcePolicy_disappears(t *testing.T) {
```
Because the underlying Pipeline could be deleted, this should also have a parent disappears test as described at https://hashicorp.github.io/terraform-provider-aws/running-and-writing-acceptance-tests/#disappears-acceptance-tests

gdavison reviewed (dismissed):
The following tests are failing:

* `TestAccOpenSearchIngestionPipelineResourcePolicy_disappears`, probably due to the empty policy case being treated as still present. The find function suggestion will likely resolve this
* `TestAccOpenSearchIngestionPipelineEndpoint_basic` and `TestAccOpenSearchIngestionPipelineEndpoint_disappears` fail with the error message `unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'`

Please update the PR description with the results of the acceptance tests

graytaylor0 commented:
> The following tests are failing:
> 
>     * `TestAccOpenSearchIngestionPipelineResourcePolicy_disappears`, probably due to the empty policy case being treated as still present. The find function suggestion will likely resolve this
> 
>     * `TestAccOpenSearchIngestionPipelineEndpoint_basic` and `TestAccOpenSearchIngestionPipelineEndpoint_disappears` fail with the error message `unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'`
> 
> 
> Please update the PR description with the results of the acceptance tests

Thanks for your patience and support. I will run the tests again and paste the full results. I ran them multiple times successfully against us-east-1. 

Are you testing against us-west-2? It seems there may be an AWS service side issue in us-west-2 that is causing these to fail as I get CREATE_FAILED as well when I run against us-west-2. Will reach out to get this fixed. Please try to test against us-east-1 to confirm they pass there if you are able to.

gdavison commented:
I get the same result in `us-east-1`: "unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'"

graytaylor0 committed:
Simplify test resources to make acceptance tests faster

Signed-off-by: Taylor Gray <tylgry@amazon.com>

graytaylor0 commented:
The OpenSearch Ingestion service team pushed a fix for the CREATE_FAILED issue, and I have been able to get both tests to pass now in us-east-1 and us-west-2

```
=== us-east-1 Results ===

--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_basic (654.86s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/osis    660.616s

--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_disappears (712.79s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/osis    720.239s

--- PASS: TestAccOpenSearchIngestionPipelineResourcePolicy_basic (477.86s)
--- PASS: TestAccOpenSearchIngestionPipelineResourcePolicy_update (944.35s)
--- PASS: TestAccOpenSearchIngestionPipelineResourcePolicy_disappears (972.51s)
--- PASS: TestAccOpenSearchIngestionPipelineResourcePolicy_disappears_pipeline (975.89s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/osis    988.193s

=== us-west-2 Results ===

--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_basic (580.04s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/osis    585.895s

--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_disappears (637.33s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/osis    644.237s
```

gdavison committed:
Allows random assignment of AZs. Prevents use of Local Zones

gdavison committed:
Removes unneeded attribute

gdavison committed:
Replaces `resource.ComposeTestCheckFunc` with `resource.ComposeAggregateTestCheckFunc`

gdavison committed:
Adds missing attribute checks and checks specific values

gdavison committed:
Tweaks documentation

gdavison committed:
Merge branch 'main' into PipelineResourcePolicy

gdavison committed:
Updates random pipeline name

gdavison committed:
Adds Resource Identity tests for `aws_osis_pipeline_resource_policy`

gdavison committed:
Adds Resource Identity to `aws_osis_pipeline_endpoint`

gdavison committed:
Re-orders `aws_osis_pipeline` documentation example

gdavison committed:
Updates import documentation for `aws_osis_pipeline_resource_policy`

gdavison committed:
Updates import documentation for `aws_osis_pipeline_endpoint`

gdavison committed:
Renames `aws_osis_pipeline_resource_policy` to `aws_osis_resource_policy` to match API struct name

gdavison committed:
Semgrep fixes

gdavison committed:
Allocates inputs on stack

gdavison committed:
Replaces `resource.ComposeTestCheckFunc` with `resource.ComposeAggregateTestCheckFunc`

gdavison committed:
Cleanup

gdavison committed:
Adds cross-account test

gdavison commented:
Acceptance test results

```console
% make testacc PKG=osis TESTS=TestAccOpenSearchIngestionPipelineEndpoint_

--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_Identity_regionOverride (522.79s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_Identity_basic (525.01s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_basic (541.40s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_disappears (579.43s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_crossAccount (606.53s)
```

```console
% make testacc PKG=osis TESTS=TestAccOpenSearchIngestionResourcePolicy_

--- PASS: TestAccOpenSearchIngestionResourcePolicy_basic (426.87s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_Identity_basic (439.13s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_Identity_regionOverride (443.34s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_disappears_pipeline (465.34s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_update (466.35s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_disappears (984.36s)
```

TODO

- [ ] List Resources for
  - [ ] `aws_osis_resource_policy`
  - [ ] `aws_osis_pipeline_endpoint`

gdavison commented:
/test

gdavison committed:
Adds `aws_osis_pipeline_endpoint` List Resource

hc-github-team-terraform-aws commented:
### Latest automated test results:

```console
% TF_ACC=1 go test './internal/service/osis/...' -count=1 -json -v -run='(TestAccOpenSearchIngestionResourcePolicy|TestAccOpenSearchIngestionPipelineEndpoint)' -parallel '20' -timeout=0 -vet=off -buildvcs=false

TestAccOpenSearchIngestionPipelineEndpoint_Identity_basic: [PASS] 679.31s
TestAccOpenSearchIngestionPipelineEndpoint_Identity_regionOverride: [PASS] 674.18s
TestAccOpenSearchIngestionPipelineEndpoint_basic: [PASS] 640.62s
TestAccOpenSearchIngestionPipelineEndpoint_disappears: [PASS] 623.12s
TestAccOpenSearchIngestionPipelineEndpoint_crossAccount: [PASS] 847.28s
TestAccOpenSearchIngestionResourcePolicy_Identity_basic: [PASS] 668.06s
TestAccOpenSearchIngestionResourcePolicy_Identity_regionOverride: [PASS] 682.06s
TestAccOpenSearchIngestionResourcePolicy_basic: [PASS] 543.9s
TestAccOpenSearchIngestionResourcePolicy_disappears: [PASS] 513.09s
TestAccOpenSearchIngestionResourcePolicy_disappears_pipeline: [PASS] 546.3s
TestAccOpenSearchIngestionResourcePolicy_update: [PASS] 563.55s
```

gdavison committed:
Tweaks List Resource documentation to mention AutoFlex options

gdavison committed:
Adds `aws_osis_resource_policy` List Resource

gdavison committed:
Updates CHANGELOG

gdavison committed:
Updates testing directions

gdavison commented:
Acceptance test results

```console
% make testacc PKG=osis TESTS=TestAccOpenSearchIngestionPipelineEndpoint_

--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_basic (504.23s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_disappears (530.62s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_List_includeResource (543.60s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_Identity_regionOverride (553.87s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_List_basic (563.03s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_crossAccount (597.57s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_Identity_basic (602.00s)
--- PASS: TestAccOpenSearchIngestionPipelineEndpoint_List_regionOverride (611.98s)
```

```console
% make testacc PKG=osis TESTS=TestAccOpenSearchIngestionResourcePolicy_

--- PASS: TestAccOpenSearchIngestionResourcePolicy_disappears_pipeline (390.82s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_basic (424.00s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_Identity_basic (456.96s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_disappears (473.52s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_update (476.16s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_Identity_regionOverride (504.08s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_List_includeResource (577.13s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_List_basic (945.69s)
--- PASS: TestAccOpenSearchIngestionResourcePolicy_List_regionOverride (994.57s)
```

gdavison commented:
/test

YakDriver reviewed (approved):
LGTM

hc-github-team-terraform-aws commented:
### Latest automated test results:

```console
% TF_ACC=1 go test './internal/service/osis/...' -count=1 -json -v -run='(TestAccOpenSearchIngestionResourcePolicy|TestAccOpenSearchIngestionPipelineEndpoint)' -parallel '20' -timeout=0 -vet=off -buildvcs=false

TestAccOpenSearchIngestionPipelineEndpoint_Identity_basic: [PASS] 729.36s
TestAccOpenSearchIngestionPipelineEndpoint_Identity_regionOverride: [PASS] 788.07s
TestAccOpenSearchIngestionPipelineEndpoint_List_basic: [PASS] 708.27s
TestAccOpenSearchIngestionPipelineEndpoint_List_includeResource: [PASS] 680.42s
TestAccOpenSearchIngestionPipelineEndpoint_List_regionOverride: [PASS] 762.61s
TestAccOpenSearchIngestionPipelineEndpoint_basic: [PASS] 700.31s
TestAccOpenSearchIngestionPipelineEndpoint_disappears: [PASS] 760.29s
TestAccOpenSearchIngestionPipelineEndpoint_crossAccount: [PASS] 831.28s
TestAccOpenSearchIngestionResourcePolicy_Identity_basic: [PASS] 1059.8s
TestAccOpenSearchIngestionResourcePolicy_Identity_regionOverride: [FAIL] 1791.06s

------- Stdout: -------
=== RUN   TestAccOpenSearchIngestionResourcePolicy_Identity_regionOverride
=== PAUSE TestAccOpenSearchIngestionResourcePolicy_Identity_regionOverride
=== CONT  TestAccOpenSearchIngestionResourcePolicy_Identity_regionOverride
    resource_policy_identity_gen_test.go:113: Step 1/6 error: Error running apply: exit status 1
        
        Error: waiting for OpenSearch Ingestion Pipeline (tf-acc-test-mo27864ugm) create
        
          with aws_osis_pipeline.test,
          on main_gen.tf line 32, in resource "aws_osis_pipeline" "test":
          32: resource "aws_osis_pipeline" "test" {
        
        unexpected state 'CREATE_FAILED', wanted target 'ACTIVE'. last error:
        OpenSearch Ingestion failed to stabilize. To get more information, try to
        delete and recreate the pipeline with log publishing enabled so you can
        troubleshoot the issue in CloudWatch Logs. If you can't resolve the issue,
        contact AWS Support.
2026/07/28 18:35:29 Creating Terraform AWS Provider (SDKv2-style)...
2026/07/28 18:35:29 Initializing Terraform AWS Provider (SDKv2-style)...
2026/07/28 18:35:29 Creating Terraform AWS Provider (Framework-style)...
2026/07/28 18:35:29 Initializing Terraform AWS Provider (Framework-style)...
2026/07/28 18:35:31 Creating Terraform AWS Provider (SDKv2-style)...
2026/07/28 18:35:31 Initializing Terraform AWS Provider (SDKv2-style)...
2026/07/28 18:35:31 Creating Terraform AWS Provider (Framework-style)...
2026/07/28 18:35:31 Initializing Terraform AWS Provider (Framework-style)...
--- FAIL: TestAccOpenSearchIngestionResourcePolicy_Identity_regionOverride (1791.06s)
TestAccOpenSearchIngestionResourcePolicy_List_basic: [PASS] 616.05s
TestAccOpenSearchIngestionResourcePolicy_List_includeResource: [PASS] 533.95s
TestAccOpenSearchIngestionResourcePolicy_List_regionOverride: [PASS] 623.29s
TestAccOpenSearchIngestionResourcePolicy_basic: [PASS] 501.06s
TestAccOpenSearchIngestionResourcePolicy_disappears: [PASS] 706.91s
TestAccOpenSearchIngestionResourcePolicy_disappears_pipeline: [PASS] 609.86s
TestAccOpenSearchIngestionResourcePolicy_update: [PASS] 981.48s
```

gdavison commented:
I'm seeing intermittent failures in creating `aws_osis_pipeline` in `us-east-1`. Not related to the changes in this PR.
